src/net/connection.cpp

```diff
@@ -47,8 +55,15 @@ void Connection::handle_write_() {
 }
 
 void Connection::handle_event_(short what) {
+    logger->debug("connection: got bufferevent event: %d", what);
 
     if (what & BEV_EVENT_EOF) {
+        auto input = bufferevent_get_input(bev);
+        if (evbuffer_get_length(input) > 0) {
+            logger->debug("Suppress EOF with data lingering in input buffer");
+            suppressed_eof = true;
+            return;
+        }
```

Comment:
I think it's ok to do this unconditionally as there may be some other edge cases we haven't yet found when the connection is closed in an unclean way. In general this logic seems like something we may always want to have, that is if we receive and EOF and there is still some data in the buffer we should process it before the EOF is triggered.